lm/lm_image.py

```diff
@@ -0,0 +1,90 @@
+__all__ = ["LmImageJob"]
+
+import shutil
+
+from sisyphus import *
+from i6_core import rasr
+from i6_core import util
+
+Path = setup_path(__package__)
+
+
+class LmImageJob(rasr.RasrCommand, Job):
+    """
+    pre-compute LM image without generating global cache
+    """
+
+    def __init__(
+        self,
+        crp,
+        extra_config=None,
+        extra_post_config=None,
+        encoding="utf-8",
+        mem=2,
+    ):
+        kwargs = locals()
+        del kwargs["self"]
+
+        self.config, self.post_config = LmImageJob.create_config(**kwargs)
+        self.exe = self.select_exe(crp.lm_util_exe, "lm-util")
+
+        self.log_file = self.log_file_output_path("lm_image", crp, False)
+        self.out_image = self.output_path("lm.image", cached=True)
+
+        self.rqmt = {"time": 1, "cpu": 1, "mem": mem}
+
+    def tasks(self):
+        yield Task("create_files", resume="create_files", mini_task=True)
+        yield Task("run", resume="run", rqmt=self.rqmt)
+
+    def create_files(self):
+        self.write_config(self.config, self.post_config, "lm_image.config")
+        extra_code = (
+            ":${{THEANO_FLAGS:="
+            "}}\n"
+            'export THEANO_FLAGS="$THEANO_FLAGS,device={0},force_device=True"\n'
+            'export TF_DEVICE="{0}"'.format(
+                "gpu" if self.rqmt.get("gpu", 0) > 0 else "cpu"
+            )
+        )
+        self.write_run_script(self.exe, "lm_image.config", extra_code=extra_code)
+
+    def run(self):
+        self.run_script(1, self.log_file)
+        shutil.move("lm.image", self.out_image.get_path())
+
+    def cleanup_before_run(self, cmd, retry, *args):
+        util.backup_if_exists("lm_image.log")
+
+    @classmethod
+    def create_config(
+        cls,
+        crp,
+        extra_config,
+        extra_post_config,
+        encoding,
+        **kwargs,
+    ):
+        config, post_config = rasr.build_config_from_mapping(
+            crp, {"lexicon": "lm-util.lexicon", "language_model": "lm-util.lm"}
+        )
+        del (
+            config.lm_util.lm.scale
+        )  # scale not considered here, delete to remove ambiguity
+
+        assert config.lm_util.lm.type == "ARPA"
+
+        config.lm_util.action = "load-lm"
+        config.lm_util.lm.image = "lm.image"
```

Comment:
You also need to check if image is set in the lm_post_config. I'm a bit undecided what to do in that case. Either keep the old image (and create a link to it as output of this job) or to recreate the image. @michelwi @Nintendo-Wei  what do you think?

Reply:
how would the image from post config be created in the first place? If I understand correctly, once this job is being used, it is the only place where images are created within the recipes. 

So if all works correctly if a second `LmImageJob` is called for the same LM, then it should be the same instance as the one that created the image and the config will be unmodified / no second image created.
In case the existing image is an external file, I would recreate it within the graph.
If the image was created by running RASR and accidentally creating it, then it will be in the work folder of the other Job and not accessible / possibly cleaned away later.
Now the only other possibility would be the image comes from the now existing `LmImageAndGlobalCacheJob` then not sure what to do.. depends on your/our plans how to handle the transition phase.

Independent of this, yes, we should check for `post_config.lm_util.lm.image` and delete it. Or maybe write lm.image in the post config in the first place. mh.. its anyway constant so doesn't matter.

Reply:
According to line 416 in `i6_core/recognition/advanced_tree_search.py`: `lm_config[1].image = ...` the image is actually saved to the post_config right? According to the definition of `find_arpa_lms`, `lm_config[0]` is config and `lm_config[1]` is post_config.
So I did a similar thing for the new PR in `apptek_asr`: https://bitbucket.org/omnifluent/apptek_asr/pull-requests/93 line 162-164.
Is that what we wanted? Please correct me if I was wrong, thanks!

Reply:
in the `find_arpa_lms` function, anyway only LMs without existing image are returned. so for the use case in `apptek_asr` it is save.

Maybe we add a warning if image already exists (this should only happen in case of user error?) and then create a new one anyway?

Reply:
Would something like 
```
if post_config.lm_util.lm.image:
    logging.warning("The LM image already exists, but a new one will be recreated.")
```
be enough?
Or we would like to delete the image file with `os.remove()`?

Reply:
It's not about how this Job would be used inside a recognition job. But about the interface of the job on it's own. A user is allowed to put the image parameter in post_config and then this job is broken. I thought about it more and I'd say the most common use case is to specifically create the LM-image. But we should then in the recognition job only call it for LM's that don't have an image yet. The job should thus be renamed `CreateLmImageJob` to better communicate this purpose.

Reply:
But most importantly: DO NOT call os.remove on the old image.

Reply:
>  I thought about it more and I'd say the most common use case is to specifically create the LM-image. 

I agree.

> But we should then in the recognition job only call it for LM's that don't have an image yet. 

yes, it is done like this in the associated PR to the `apptek_asr` recipes

> The job should thus be renamed CreateLmImageJob to better communicate this purpose.

ok, we can do it

> DO NOT call os.remove on the old image.

yes!